x/hold/simulation/genesis.go

```diff
@@ -0,0 +1,103 @@
+package simulation
+
+import (
+	"fmt"
+	"math/rand"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	simtypes "github.com/cosmos/cosmos-sdk/types/simulation"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	hold "github.com/provenance-io/provenance/x/hold"
+)
+
+const HoldAccountHolds = "hold-account-holds"
+
+// RandomAccountHolds randomly selects accounts with an existing balance to place a hold of a random amount.
+func RandomAccountHolds(r *rand.Rand, accounts []simtypes.Account) []*hold.AccountHold {
+	if len(accounts) == 0 {
+		return nil
+	}
+
+	count := r.Intn(len(accounts) + 1)
+	if count == 0 {
+		return nil
+	}
+
+	addrs := make([]sdk.AccAddress, len(accounts))
+	for i, acc := range accounts {
+		addrs[i] = acc.Address
+	}
+	r.Shuffle(len(addrs), func(i, j int) {
+		addrs[i], addrs[j] = addrs[j], addrs[i]
+	})
+
+	rv := make([]*hold.AccountHold, count)
+	for i, addr := range addrs[:count] {
+		rv[i] = &hold.AccountHold{
+			Address: addr.String(),
+			Amount:  sdk.NewCoins(sdk.NewInt64Coin(sdk.DefaultBondDenom, r.Int63n(1000)+1)),
+		}
+	}
+
+	return rv
+}
+
+// UpdateBankGenStateForHolds adds all hold funds to the bank balances.
+// Panics if there's an address with a hold that doesn't already have a balance.
+func UpdateBankGenStateForHolds(bankGenState *banktypes.GenesisState, holdGenState *hold.GenesisState) {
+	if len(holdGenState.Holds) == 0 {
+		return
+	}
+
+	var totalAdded sdk.Coins
+HoldsLoop:
+	for _, ah := range holdGenState.Holds {
+		for i, bal := range bankGenState.Balances {
+			if ah.Address == bal.Address {
+				bankGenState.Balances[i].Coins = bal.Coins.Add(ah.Amount...)
+				totalAdded = totalAdded.Add(ah.Amount...)
+				continue HoldsLoop
+			}
+		}
+		panic(fmt.Errorf("no bank genesis balance found for %s that should have a hold on %s", ah.Address, ah.Amount))
+	}
+
+	bankGenState.Supply = bankGenState.Supply.Add(totalAdded...)
+}
+
+// holdsString creates a JSON object string of address -> amount for each hold.
+func holdsString(holds []*hold.AccountHold) string {
+	if len(holds) == 0 {
+		return "{}"
+	}
+	lines := make([]string, len(holds))
+	for i, ah := range holds {
+		lines[i] = fmt.Sprintf("%q:%q", ah.Address, ah.Amount)
+	}
+	return fmt.Sprintf("{\n %s\n}", strings.Join(lines, ",\n "))
+}
+
+// RandomizedGenState generates a random GenesisState for the hold module.
+func RandomizedGenState(simState *module.SimulationState) {
+	holdGenState := hold.DefaultGenesisState()
+
+	simState.AppParams.GetOrGenerate(
+		simState.Cdc, HoldAccountHolds, &holdGenState.Holds, simState.Rand,
+		func(r *rand.Rand) {
+			holdGenState.Holds = RandomAccountHolds(r, simState.Accounts)
```

Comment:
Will any of the sims generate a bad account address? If not, it may be helpful to sometimes create a bad one to sim test validation.

Reply:
They'll all be valid. Introducing an invalid one would actually cause the sim to fail. There's no way to communicate that you're introducing something that's expected to error during genesis. And even if there were, such an error would then prevent any of the rest of the sim from doing anything.